source/extensions/filters/http/jwt_authn/verifier.cc

```diff
@@ -0,0 +1,228 @@
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRequirement;
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRequirementAndList;
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRequirementOrList;
+using ::google::jwt_verify::Status;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace JwtAuthn {
+
+// base verifier for provider_name, provider_and_audiences, and allow_missing_or_failed.
+class BaseVerifierImpl : public Verifier {
+public:
+  BaseVerifierImpl(const BaseVerifierImpl* parent) : parent_(parent) {}
+
+  void onCompleteHelper(Status status, VerifyContext& context) const {
+    if (parent_ != nullptr) {
+      return parent_->onComplete(status, context);
+    }
+    return context.callback()->onComplete(status);
+  }
+
+  // Check if next verifier should be notified of status, or if no next verifier exists signal
+  // callback in context.
+  virtual void onComplete(const Status& status, VerifyContext& context) const {
+    auto& response_data = context.getResponseData(this);
+    if (!response_data.has_responded_) {
+      response_data.has_responded_ = true;
+      onCompleteHelper(status, context);
+    }
+  }
+
+protected:
+  // The parent group verifier.
+  const BaseVerifierImpl* const parent_;
+};
+
+VerifierPtr innerCreate(const JwtRequirement& requirement,
+                        const Protobuf::Map<ProtobufTypes::String, JwtProvider>& providers,
+                        const AuthFactory& factory, const Extractor& extractor,
+                        const BaseVerifierImpl* parent);
+
+namespace {
+
+// Provider specific verifier
+class ProviderVerifierImpl : public BaseVerifierImpl {
+public:
+  ProviderVerifierImpl(const std::vector<std::string>& audiences, const AuthFactory& factory,
+                       const JwtProvider& provider, const BaseVerifierImpl* parent)
+      : BaseVerifierImpl(parent), audiences_(audiences), auth_factory_(factory),
+        extractor_(
+            Extractor::create(provider.issuer(), provider.from_headers(), provider.from_params())),
+        issuer_(provider.issuer()) {}
+
+  void verify(VerifyContext& context) const override {
+    auto auth = auth_factory_.create(audiences_, absl::optional<std::string>{issuer_}, false);
+    auth->sanitizePayloadHeaders(context.headers());
+    auth->verify(context.headers(), extractor_->extract(context.headers()),
+                 [&](const Status& status) { onComplete(status, context); });
+    if (!context.getResponseData(this).has_responded_) {
+      context.addAuth(std::move(auth));
+    }
+  }
+
+private:
+  const std::vector<std::string> audiences_;
+  const AuthFactory& auth_factory_;
+  const ExtractorConstPtr extractor_;
+  const std::string issuer_;
+};
+
+// Allow missing or failed verifier
+class AllowFailedVerifierImpl : public BaseVerifierImpl {
+public:
+  AllowFailedVerifierImpl(const AuthFactory& factory, const Extractor& extractor, bool allow_failed,
```

Comment:
I believe pass_in extractor is only needed for this "allow_fail" case.   It will be eaiser to read if we name it as all_extractor or default_extractor.